src/gui/createGrapheneSpeciesDialog.cpp

```diff
@@ -0,0 +1,521 @@
+
+    // Set display configuration
+    ui_.StructureViewer->setSpecies(&species_);
+
+    refreshLock.unlock();
+
+    calculateParameters();
+
+    regenerate();
+}
+
+/*
+ * Data
+ */
+
+// Calculate parameters
+void CreateGrapheneSpeciesDialog::calculateParameters()
+{
+    /*
+     * Procedure based on "Determination of the chiral indices (n,m) of carbon nanotubes by electron diffraction",
+     * Lu-Chan Qin, Phys. Chem. Chem. Phys. 2007, *9*, 31-48. https://doi.org/10.1039/B614121H
+     */
+
+    // Get n,m values for tube
+    n_ = ui_.NSpin->value();
+    m_ = ui_.MSpin->value();
+
+    // Determine a0, the principal unit length of the structure, from the specified bond distance. See Figure 1a, noting that
+    // a1 == a2 == a0.
+    const auto r = ui_.BondLengthSpin->value();
+    a0_ = r * root3;
+
+    // Set principal vectors for reference
+    va1_ = Vec3<double>(a0_, 0.0, 0.0);
+    va2_ = Vec3<double>(a0_ * sin(oneSixthPi), a0_ * cos(oneSixthPi), 0.0);
+
+    // Calculate repeat dimensions of the (n,m) sheet
+    // -- A is magnitude of vecA (equation 1) and represents our X dimension
+    A_ = a0_ * sqrt(n_ * n_ + m_ * m_ + n_ * m_);
+    // -- c is the magnitude of the vector perpendicular to vecA, along the tube axis, and represents our Y dimension
+    const auto M = std::gcd(2 * n_ + m_, n_ + 2 * m_);
+    c_ = sqrt(3.0) * A_ / M;
+
+    // Determine angle alpha, describing the tilt of the sheet away from va1 (equation 3)
+    alpha_ = atan((sqrt(3.0) * m_) / (2 * n_ + m_));
+
+    // The radius of the tube is then the circumference (== A) divided by 2PI
+    radius_ = A_ / (2.0 * M_PI);
+
+    dy_ = va2_.y / cos(alpha_);
+    H_ = round(c_ / dy_);
+}
+
+// Find dangling atoms, defined as those which have two bonds spanning PBC of the box
+std::map<SpeciesAtom *, std::vector<SpeciesAtom *>> CreateGrapheneSpeciesDialog::findDanglingAtoms(double localCutoff)
+{
+    std::map<SpeciesAtom *, std::vector<SpeciesAtom *>> atoms;
+    for (auto &i : species_.atoms())
+    {
+        // Find all bound partners which are across the unit cell
+        std::vector<SpeciesAtom *> pbcJ;
+        for (auto &b : i.bonds())
+        {
+            auto *j = b.get().partner(&i);
+            if ((i.r() - j->r()).magnitude() > localCutoff)
+                pbcJ.push_back(j);
+        }
+
+        if (pbcJ.size() == 2)
+            atoms[&i] = pbcJ;
+    }
+
+    return atoms;
+}
+
+// Get vector of atom keys from atom/neighbour map, sorted by position along indicated direction
+std::vector<SpeciesAtom *>
+CreateGrapheneSpeciesDialog::getSorted(const std::map<SpeciesAtom *, std::vector<SpeciesAtom *>> &atoms, int dir) const
+{
+    std::vector<SpeciesAtom *> sorted;
+    for (auto &&[i, _] : atoms)
+        sorted.push_back(i);
+
+    std::sort(sorted.begin(), sorted.end(),
+              [dir](const SpeciesAtom *a, const SpeciesAtom *b) { return a->r().get(dir) < b->r().get(dir); });
+
+    return sorted;
+}
+
```

Comment:
Small thing, but, since this is only ever used in one place, we could just make this `getSortedByY`

```suggestion
// Get vector of atom keys from atom/neighbour map, sorted by position along indicated direction
std::vector<SpeciesAtom *>
CreateGrapheneSpeciesDialog::getSortedByY(const std::map<SpeciesAtom *, std::vector<SpeciesAtom *>> &atoms) const
{
    std::vector<SpeciesAtom *> sorted;
    for (auto &&[i, _] : atoms)
        sorted.push_back(i);

    std::sort(sorted.begin(), sorted.end(),
              [dir](const SpeciesAtom *a, const SpeciesAtom *b) { return a->r().y < b->r().y; });

    return sorted;
}

```

Reply:
Ah, yes.  This was introduced thinking there would be more trials of sorting along different directions, but as it turned out not so much!